### PR TITLE
multiple keyboards: PORT_CHAR/PORT_CODE fixes for Lua (part 1)

### PR DESCRIPTION
**Pull request by shattered**
Allows emu.keypost() in relevant drivers to input entire ASCII table correctly.

### CODE REVIEW

**Comment by happppp** [comment]
PORT_CHAR(0 to 25) is same as PORT_CHAR('A' to 'Z')??
That sounds more like emu.keypost has a bug, rather than the input defs in
drivers.

**Comment by ajrhacker** [comment]
Those are the control-modified key values.

**Comment by cuavas** [comment]
I'll check this later today - I'm in touch with @shattered

File: src/devices/bus/pc_kbd/ec1841.cpp [hunk 1]
```diff
@@ -242,7 +242,7 @@ INPUT_PORTS_START( ec_1841_keyboard )
 
 	PORT_START("MD14")
 	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_CODE(KEYCODE_BACKSLASH) PORT_CHAR('\\') PORT_CHAR('|')
```
Comment (cuavas) [inline src/devices/bus/pc_kbd/ec1841.cpp]: Does this key not produce ^\ (0x1C) with control held?

File: src/devices/bus/pc_kbd/ec1841.cpp [hunk 2]
```diff
@@ -202,7 +202,7 @@ INPUT_PORTS_START( ec_1841_keyboard )
 
 	PORT_START("MD10")
 	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_CODE(KEYCODE_9) PORT_CHAR('9') PORT_CHAR('(')
-	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_CODE(KEYCODE_P) PORT_CHAR('p') PORT_CHAR('P')
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_CODE(KEYCODE_P) PORT_CHAR('p') PORT_CHAR('P') PORT_CHAR(0x10)
 	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_CODE(KEYCODE_CLOSEBRACE) PORT_CHAR(']') PORT_CHAR('}')
```
Comment (cuavas) [inline src/devices/bus/pc_kbd/ec1841.cpp]: Does this key not generate ^] (0x1D) with control held?

File: src/mame/machine/ibm6580_kbd.cpp [hunk 3]
```diff
@@ -89,69 +89,69 @@ INPUT_PORTS_START( dw_keyboard )
 
 	PORT_START("COL.3")
 	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("/ ?") PORT_CODE(KEYCODE_SLASH) PORT_CHAR('/') PORT_CHAR('?')
-	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("3 2") PORT_CODE(KEYCODE_BACKSLASH)
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("3 2") PORT_CODE(KEYCODE_BACKSLASH) PORT_CHAR('\\') PORT_CHAR('|')
 	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("[ ]") PORT_CODE(KEYCODE_CLOSEBRACE) PORT_CHAR('[') PORT_CHAR(']')
 	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_EQUALS) PORT_CHAR('=') PORT_CHAR('+')
 	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_MINUS) PORT_CHAR('-') PORT_CHAR('_')
-	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("1/4 1/2") PORT_CODE(KEYCODE_OPENBRACE)
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("1/4 1/2") PORT_CODE(KEYCODE_OPENBRACE) PORT_CHAR('[') PORT_CHAR('{')
```
Comment (cuavas) [inline src/mame/machine/ibm6580_kbd.cpp]: If this key is for the the 1/4 and 1/2 fractions, you should use 0xbc and 0xbd as the `PORT_CHAR` codes, not artificially map it to something unrelated.

File: src/mame/machine/ibm6580_kbd.cpp [hunk 4]
```diff
@@ -89,69 +89,69 @@ INPUT_PORTS_START( dw_keyboard )
 
 	PORT_START("COL.3")
 	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("/ ?") PORT_CODE(KEYCODE_SLASH) PORT_CHAR('/') PORT_CHAR('?')
-	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("3 2") PORT_CODE(KEYCODE_BACKSLASH)
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("3 2") PORT_CODE(KEYCODE_BACKSLASH) PORT_CHAR('\\') PORT_CHAR('|')
```
Comment (cuavas) [inline src/mame/machine/ibm6580_kbd.cpp]: Is this for superscript 3 and 2?  If it is, the `PORT_CHAR` should be 0xb2 and 0xb3, not something unrelated.

File: src/mame/machine/ibm6580_kbd.cpp [hunk 5]
```diff
@@ -89,69 +89,69 @@ INPUT_PORTS_START( dw_keyboard )
...
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_Q) PORT_CHAR('q') PORT_CHAR('Q') PORT_CHAR(0x11)
+	PORT_BIT( 0x40, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_A) PORT_CHAR('a') PORT_CHAR('A') PORT_CHAR(0x01)
+	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_Z) PORT_CHAR('z') PORT_CHAR('Z') PORT_CHAR(0x1a)
 
 	PORT_START("COL.9")
 	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("LShift") PORT_CODE(KEYCODE_LSHIFT) PORT_CHAR(UCHAR_SHIFT_1)
 	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("Lock") PORT_CODE(KEYCODE_CAPSLOCK) PORT_CHAR(UCHAR_MAMEKEY(CAPSLOCK))
 	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_TAB) PORT_CHAR('\t')
-	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("Plusminus") PORT_CODE(KEYCODE_TILDE)
+	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("Plusminus") PORT_CODE(KEYCODE_TILDE) PORT_CHAR('`') PORT_CHAR('~')
```
Comment (cuavas) [inline src/mame/machine/ibm6580_kbd.cpp]: Please use 0xb1 for this character, and whatever it should be for the shifted version of the key, not unrelated characters.

Comment (ajrhacker) [inline src/mame/machine/ibm6580_kbd.cpp]: The shifted version looks like the degree symbol (0xb0).

**Comment by shattered** [comment]
> Mapping keys to unrelated characters defeats the purpose of natural keyboard - it's easy enough for me to ± (Option-Shift-Equals) and that should map to the ± key, I shouldn't have to type some other character in natural mode, and the same goes for other characters

I don't understand this comment.   Perhaps I don't understand what natural keyboard does, either...

Anyway, a version of this PR that only touches Soviet keyboards is https://github.com/mamedev/mame/pull/2790.

**Comment by cuavas** [comment]
The point of "natural keyboard" mode and `PORT_CHAR` is to tell MAME what the character is typed by a key/combination so that MAME can translate from character events on the host system to keystrokes on the emulated system.

Consider a an emulated system with a JIS "bitshift" keyboard layout and a host system with a US "typewriter shift" layout.  On the emulated system, the key to the right of P has the characters "@" (at sign) and "`" (backtick) on it.  The `PORT_CHAR` macros for this key should use these characters, so that when the user types Shift-2 on the host system posting an "@" character, it's translated to pressing and releasing this single key, or when they type the unshifted "`" key (to the left of 1), it's translated to pressing shift, pressing the "@" key, releasing the "@" key, and releasing shift.

Or suppose the emulated system has a German QWERTZ layout.  There's a "Ü" key to the right of P on the emulated system in this case.  How the user types "ü" depend on their keyboard layout: with Windows US International, it's Shift-" u; with Mac US it's Option-U u; with a German layout on the host system, it's the Ü key to the right of P.  In all these cases, MAME needs to translate the character event to a single emulated keystroke.

Likewise, it needs to automatically press shift to type numbers on French AZERTY, adjust Z/Y for German QWERTZ, remap £/#/@/" for UK layouts, etc.

The lua keypost function piggybacks on the natural keyboard features and posts strings of characters that MAME has to translate to keystrokes using this mechanism.  The characters posted should correspond to the characters typed on the emulated system.

If there's a key that types "±" on the emulated system, the `PORT_CHAR` macro should have the Unicode for ± (0x00b1) so typing/posting "±" (Option-Shit-= on Mac US layout, or Alt+0177 on Windows) behaves "naturally".  It should not have an unrelated character that happens to be in the same position on a different keyboard layout.

**Comment by shattered** [comment]
Will regenerate the PR

**Pull request closed without merging**